src/field.rs

```diff
@@ -91,6 +91,52 @@ pub trait Field:
         let int: Self::Integer = (*self).into();
         int.into()
     }
+
+    /// Generic implementation to serialize fields into a buffer. Callers need to make sure
+    /// there is enough capacity to store the value of this field.
+    /// It is less efficient because it operates with generic representation of fields as 16 byte
+    /// integers, so consider overriding it for actual field implementations
+    ///
+    /// ## Errors
+    /// Returns an error if buffer did not have enough capacity to store this field value
+    fn serialize<W: Write>(&self, writer: &mut W) -> io::Result<()> {
+        let bytes = Self::Integer::BITS as usize / 8;
+        let raw_value = &self.as_u128().to_le_bytes()[..bytes];
+
+        let expected_capacity = raw_value.len();
+        let actually_written = writer.write(raw_value)?;
+        if actually_written == raw_value.len() {
+            Ok(())
+        } else {
+            let error_text = format!(
+                "Buffer does not have enough capacity to hold \
+                {expected_capacity} bytes needed to store the field value {self:?}. Only \
+                {actually_written} bytes have been written to it."
+            );
+
+            Err(io::Error::new(ErrorKind::WriteZero, error_text))
+        }
+    }
+
+    /// Generic implementation to deserialize fields from buffer.
+    /// It is less efficient because it allocates 16 bytes on the stack to accommodate for all
+    /// possible field implementations, so consider overriding it for actual field implementations
+    ///
+    /// In the bright future when we have const generic expressions, this can be changed to provide
+    /// zero-cost generic implementation
+    ///
+    /// ## Errors
+    /// Returns an error if buffer did not have enough capacity left to read the field value.
+    fn deserialize<R: Read>(reader: &mut R) -> io::Result<Self> {
```

Comment:
Is there a reason you are using this specific api? You aren’t implementing a trait so this could be as simple as taking a slice (mut for write) and panicking if something goes wrong ( as it really shouldn’t)

Reply:
that's right and I started with taking `&mut [u8]` (in fact #150 has it). The thing that I didn't like (and maybe I am wrong) is that writer needs to know the offset to write the value. If we were to re-use a single slice to write two values for instance, something needs to take care of providing two different slices or make `offset` an argument to the API. 

but if our API has the shape of `&mut [u8], offset: usize`, then a typical `Write`/`Read` implementation takes care of offset tracking nicely.

I don't have strong feelings about it, if you think there is a good reason to take `&mut [u8]` instead, I can do that

Reply:
I'm just concerned that we will need to do out-of-order writes, which aren't easy to do with a writer.

(For byte-level placements, we don't need an offset argument as we can just pass in the slice that we want the field to write into.  We only need an offset if there is a need to do bit-level placement.  Avoid that for now.)

Reply:
let me make that change then - I'd love to learn more about out-of-order writes later. Is it because we need to convert BE to LE?

Reply:
done